db/redshift.js

```diff
@@ -4,251 +4,174 @@
-                await copyDataFromS3(stagingTable, tableName)
-
-                // Merge data from temporary table into the permanent table
-                await mergeDataIntoPermanentTable(stagingTable, tableName, primaryKey)
-
-                // Drop the temporary table
-                await dropTable(stagingTable)
-
-                await commit()
-                resolve()
-              } catch (error) {
-                throw new Error('Failed to get data into Redshift: ' + error)
-              }
-            })
-          }
-        })
-      })
+  /**
+   * Copy data from postgres to s3
+   * @param {String} table
+   * @param {String} s3Key
+   * @returns {Promise<boolean>} true if data was copied, false otherwise
+   */
+  const copyToS3 = async (table, s3Key) => {
+    const lastSuccess = await getLastSuccess(table)
+    const QUERY =
+      `COPY (SELECT * FROM ${table} WHERE updated_at >= '${lastSuccess}') TO STDOUT WITH(FORMAT CSV, HEADER false);`
+    const postgresClient = new Client({
+      user: process.env.DB_ENV_POSTGRESQL_USER,
+      password: process.env.DB_ENV_POSTGRESQL_PASS,
+      database: process.env.DB_ENV_POSTGRESQL_DB,
+      host: process.env.DB_PORT_5432_TCP_ADDR || 'localhost',
+      port: process.env.DB_PORT_5432_TCP_PORT || 5432
     })
+    await postgresClient.connect()
+    try {
+      let body = await postgresClient.query(copyToSteam(QUERY)).pipe(zlib.createGzip())
+      await s3.upload({
+        Bucket: process.env.REDSHIFT_S3_BUCKET,
+        Key: s3Key,
+        Body: body
+      }).promise()
+      return body.bytesRead !== 0
+    } finally {
+      // Always disconnect from postgres
+      await postgresClient.end()
+    }
   }
 
-  const updateLastSuccess = async () => {
-    return new Promise((resolve, reject) => {
-      redisClient.on('error', (error) => {
-        throw new Error('Error connecting to Redis: ' + error)
-      })
-
-      redisClient.set(LAST_SUCCESS_KEY, new Date(Date.now()).toISOString())
-      resolve()
+  /**
+   * Copy data from S3 into redshift
+   * @param {String} table
+   * @param {String} idColumn
+   * @param {String} s3Key
+   * @returns {Promise<void>}
+   */
+  const copyToRedshift = async (table, idColumn, s3Key) => {
+    const CREATE_TABLE_QUERY =
+      `CREATE TEMP TABLE ${STAGING_PREFIX}${table} (LIKE scale_of_belief.${table})`
+    const COPY_FROM_QUERY =
+      `COPY ${STAGING_PREFIX}${table}
+      FROM 's3://${process.env.REDSHIFT_S3_BUCKET}/${s3Key}'
+      IAM_ROLE '${process.env.REDSHIFT_IAM_ROLE}'
+      CSV BLANKSASNULL EMPTYASNULL TRUNCATECOLUMNS COMPUPDATE OFF STATUPDATE OFF GZIP`
+    const DELETE_QUERY =
+      `DELETE FROM scale_of_belief.${table}
+      USING ${STAGING_PREFIX}${table}
+      WHERE scale_of_belief.${table}.${idColumn} = ${STAGING_PREFIX}${table}.${idColumn}`
+    const INSERT_QUERY =
+      `INSERT INTO scale_of_belief.${table} SELECT * FROM ${STAGING_PREFIX}${table}`
```

Comment:
I realize this comment is unrelated to this PR (other than the fact that this line got refactored): This would be more efficiently done with an APPEND statement:
```
ALTER TABLE scale_of_belief.${table} APPEND FROM ${STAGING_PREFIX}${table}
```

Reply:
I believe that would add duplicates since redshift doesn't enforce unique index.

Reply:
It should bring over the exact same data as an `INSERT INTO ... SELECT *` since you don't have a `WHERE` clause on your `SELECT`

Reply:
https://docs.aws.amazon.com/redshift/latest/dg/r_ALTER_TABLE_APPEND.html